internal/impl/mysql/input_mysql_stream.go

```diff
@@ -0,0 +1,645 @@
+				}
+
+				for numRowsProcessed < rowsCount {
+					var batchRows *sql.Rows
+					if numRowsProcessed == 0 {
+						batchRows, err = i.snapshot.querySnapshotTable(table, tablePks, nil, i.fieldSnapshotMaxBatchSize)
+						if err != nil {
+							return err
+						}
+					} else {
+						batchRows, err = i.snapshot.querySnapshotTable(table, tablePks, &lastSeenPksValues, i.fieldSnapshotMaxBatchSize)
+						if err != nil {
+							return err
+						}
+					}
+
+					for batchRows.Next() {
+						numRowsProcessed++
+
+						columns, err := batchRows.Columns()
+						if err != nil {
+							batchRows.Close()
+							return err
+						}
+
+						values := make([]interface{}, len(columns))
+						valuePtrs := make([]interface{}, len(columns))
+						for i := range values {
+							valuePtrs[i] = &values[i]
+						}
+
+						if err := batchRows.Scan(valuePtrs...); err != nil {
+							batchRows.Close()
+							return err
+						}
+
+						row := map[string]any{}
+						for idx, value := range values {
+							row[columns[idx]] = value
+							if _, ok := lastSeenPksValues[columns[idx]]; ok {
+								lastSeenPksValues[columns[idx]] = value
+							}
+						}
+
+						// build message
+						i.rawMessageEvents <- MessageEvent{
+							Row:       row,
+							Operation: MessageOperationInsert,
+							Type:      MessageTypeSnapshot,
+							Table:     table,
+							Position:  nil,
+						}
+					}
+				}
+
+				return nil
+			})
+		}
+
+		if err = wg.Wait(); err != nil {
+			i.errors <- fmt.Errorf("snapshot processing failed: %w", err)
+		}
+
+		i.logger.Infof("Snapshot is done...Running CDC from BingLog: %s  on pos: %d", startPos.Name, startPos.Pos)
+		if err := i.canal.RunFrom(*startPos); err != nil {
+			i.errors <- fmt.Errorf("failed to start streaming: %v", err)
+		}
+	} else {
+		coords, _ := i.canal.GetMasterPos()
+		// starting from the last stored binlog position
+		if i.startBinLogPosition != nil {
+			coords = *i.startBinLogPosition
+		}
+
+		i.currentLogPosition = &coords
+		if err := i.canal.RunFrom(coords); err != nil {
+			i.errors <- fmt.Errorf("failed to start streaming: %v", err)
+		}
+	}
+}
+
+func (i *mysqlStreamInput) flushBatch(ctx context.Context, checkpointer *checkpoint.Capped[*int64], batch service.MessageBatch, binLogPos *mysqlReplications.Position) bool {
+	if batch == nil {
+		return true
+	}
+
+	var intPos *int64
+	if binLogPos != nil {
+		posInInt := int64(binLogPos.Pos)
+		intPos = &posInInt
+	}
+
+	resolveFn, err := checkpointer.Track(ctx, intPos, int64(len(batch)))
+	if err != nil {
+		if ctx.Err() == nil {
+			i.logger.Errorf("Failed to checkpoint offset: %v\n", err)
+		}
+		return false
+	}
+
+	lastMsg := batch[len(batch)-1]
+
+	select {
+	case i.msgChan <- asyncMessage{
+		msg: batch,
+		ackFn: func(ctx context.Context, res error) error {
+			maxOffset := resolveFn()
+			if maxOffset == nil {
+				return nil
+			}
+
+			if msgType, ok := lastMsg.MetaGet("type"); ok && msgType == "snapshot" {
```

Comment:
This is wrong. Remember the resolveFn may give back an offset that isn't tied to this specific message. So it's possible to process 100 stream messages *then* the snapshot message is ack'd which means that we should save that offset from the 100th message not this snapshot message.

Reply:
This is still incorrect.